File: wrappers/js/examples/account.js
// Account creation and key derivation (offline).
//
// Run from wrappers/js:
//
//   LIB_DIR=../../core/build/native/nativeCompile
//   CCL_LIB_PATH=$LIB_DIR DYLD_LIBRARY_PATH=$LIB_DIR LD_LIBRARY_PATH=$LIB_DIR \
//     bun examples/account.js
import { CclBridge, TESTNET } from '../src/index.js';

const bridge = new CclBridge();
try {
  // 1. Create a brand-new testnet account (random mnemonic).
  const account = bridge.account.create(TESTNET);
  const { mnemonic, base_address } = account;
  console.log('Created account');
  console.log('  base address:', base_address);
  console.log('  mnemonic    :', mnemonic);

  // 2. Restore the same account from its mnemonic — the address must match.
  const restored = bridge.account.fromMnemonic(mnemonic, TESTNET, 0, 0);
  if (restored.base_address !== base_address) throw new Error('address mismatch');
  console.log('Restored from mnemonic — address matches:', restored.base_address);

  // 3. Derive keys.
  console.log('  private key (extended, hex):', bridge.account.getPrivateKey(mnemonic, TESTNET));
  console.log('  public key (hex)           :', bridge.account.getPublicKey(mnemonic, TESTNET));

  // 4. Derive the governance DRep ID.
  console.log('  DRep ID:', bridge.account.getDrepId(mnemonic, TESTNET));
} finally {
  bridge.close();
}

File: wrappers/rust/examples/account.rs
//! Account creation and key derivation (offline).
//!
//! Run from wrappers/rust:
//!
//! ```text
//! LIB_DIR=../../core/build/native/nativeCompile
//! CCL_LIB_PATH=$LIB_DIR DYLD_LIBRARY_PATH=$LIB_DIR LD_LIBRARY_PATH=$LIB_DIR \
//!   cargo run --example account
//! ```
use ccl::{Bridge, Network};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bridge = Bridge::new()?;

    // 1. Create a brand-new testnet account (random mnemonic). Methods return JSON.
    let created = bridge.account().create(Network::Testnet)?;
    let account: serde_json::Value = serde_json::from_str(&created)?;
    let mnemonic = account["mnemonic"].as_str().unwrap();
    let base_address = account["base_address"].as_str().unwrap();
    println!("Created account");
    println!("  base address: {}", base_address);
    println!("  mnemonic    : {}", mnemonic);

    // 2. Restore the same account from its mnemonic — the address must match.
    let restored = bridge.account().from_mnemonic(mnemonic, Network::Testnet, 0, 0)?;
    let restored: serde_json::Value = serde_json::from_str(&restored)?;
    assert_eq!(restored["base_address"].as_str().unwrap(), base_address);
    println!("Restored from mnemonic — address matches: {}", base_address);

    // 3. Derive keys.
    let priv_key = bridge.account().get_private_key(mnemonic, Network::Testnet, 0, 0)?;
    let pub_key = bridge.account().get_public_key(mnemonic, Network::Testnet, 0, 0)?;
    println!("  private key (extended, hex): {}", priv_key);
    println!("  public key (hex)           : {}", pub_key);

    // 4. Derive the governance DRep ID.
    let drep_id = bridge.account().get_drep_id(mnemonic, Network::Testnet, 0)?;
    println!("  DRep ID: {}", drep_id);
    Ok(())
}
